File: algorithms/sorting/quick_sort.py
from utils.visualization_helpers import visualize_array

def quick_sort(arr, canvas, step_by_step=False):
    def partition(low, high):
        pivot = arr[(low + high) // 2]
        i = low - 1
        j = high + 1
        while True:
            i += 1
            while arr[i] < pivot:
                i += 1
            j -= 1
            while arr[j] > pivot:
                j -= 1
            if i >= j:
                return j
            arr[i], arr[j] = arr[j], arr[i]
            visualize_array(arr, canvas)
            if step_by_step:
                input("Press Enter to continue...")

    def _quick_sort(low, high):
        if low < high:
            split_index = partition(low, high)
            _quick_sort(low, split_index)
            _quick_sort(split_index + 1, high)

    _quick_sort(0, len(arr) - 1)
    return arr

File: src/algorithms/sorting/quickSort.ts
interface Step {
  array: number[];
  comparing: number[];
  swapping: boolean;
}

export function quickSort(arr: number[]): Step[] {
  const steps: Step[] = [];
  const array = [...arr];

  function partition(low: number, high: number): number {
    const pivot = array[high];
    let i = low - 1;

    for (let j = low; j < high; j++) {
      steps.push({ array: [...array], comparing: [j, high], swapping: false });
      if (array[j] < pivot) {
        i++;
        [array[i], array[j]] = [array[j], array[i]];
        steps.push({ array: [...array], comparing: [i, j], swapping: true });
      }
    }

    [array[i + 1], array[high]] = [array[high], array[i + 1]];
    steps.push({ array: [...array], comparing: [i + 1, high], swapping: true });

    return i + 1;
  }

  function quickSortHelper(low: number, high: number) {
    if (low < high) {
      const pi = partition(low, high);
      quickSortHelper(low, pi - 1);
      quickSortHelper(pi + 1, high);
    }
  }

  quickSortHelper(0, array.length - 1);
  steps.push({ array: [...array], comparing: [], swapping: false });
  return steps;
}